core/HoistAuthModel.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2024 Extremely Heavy Industries Inc.
+ */
+import {HoistModel, PlainObject, XH} from './';
+
+/**
+ *  Base class for managing authentication lifecycle.
+ *
+ *  Hoist will consult this model early in  the initialization sequence, prior to
+ *  full Hoist initialization. This means that several core services (identity, configs, prefs)
+ *  will *not* be available, but it provides the app a hook to
+ *  do early service initialization or other work to support flows such as OAuth.
+ *
+ *  The base implementation of this class is minimal and would be adequate only for fully
+ *  transparent SSO based solutions such as NTLM.  Applications wishing to provide custom
+ *  authentication should provide an extended version of this class to `XH.renderApp`
+ */
+export class HoistAuthModel extends HoistModel {
+    /**
+     * Main Entry Point
+     *
+     * This method will be called *once* by Hoist early in the app loading lifecycle to determine
+     * if full app loading can proceed.
+     *
+     * It should return true if the user could be successfully authenticated, false if the user
+     * is not authenticated, and throw in the case of technical failure. If false is returned and
+     * `AppSpec.enableLoginForm` is true the framework will enter the 'loginRequired' state and
+     * show the user a *manual* login form.
+     *
+     *  The default implementation of this method simply checks the auth status on the server, which
+     *  may be appropriate for fully SSO (e.g. NTLM) based solutions.  Override to consult or
+     *  initialize third-party client resources such as OAuth.
+     */
+    async completeAuthAsync(): Promise<boolean> {
+        return this.getAuthStatusFromServerAsync();
+    }
+
+    /**
+     * Confirm that this user has an authenticated session with the server.
+     */
+    async getAuthStatusFromServerAsync(): Promise<boolean> {
+        return XH.fetchService
+            .fetchJson({url: 'xh/authStatus'})
+            .then(r => r.authenticated)
+            .catch(e => {
+                if (e.httpStatus === 401) return false;
+                throw e;
+            });
+    }
+
+    /**
+     * Process a manual login, submitted by user via form.
+     *
+     * Return true if user is successfully logged in, otherwise false.
+     */
+    async loginFromFormAsync(username: string, password: string): Promise<boolean> {
```

Comment:
Minor thing that I can't explain, but my brain is short-circuiting on the 'Form' + 'Form' combination :) Maybe rename this to e.g. `loginWithCredentials()`?

Reply:
good suggestion, agree